components/homme/src/share/cxx/SphereOperators.hpp

```diff
@@ -1176,6 +1176,273 @@ class SphereOperators
   Real m_scale_factor_inv, m_laplacian_rigid_factor;
 };
 
+using TeamPolicy = Kokkos::TeamPolicy<ExecSpace>;
+using Team = TeamPolicy::member_type;
+
+static constexpr int NPNP = NP * NP;
+static constexpr int WARP_SIZE = warpSize;
+
+static constexpr int SPHERE_BLOCK_LEV = WARP_SIZE;
+static constexpr int SPHERE_BLOCKS_PER_COL = (NUM_PHYSICAL_LEV - 1) / SPHERE_BLOCK_LEV + 1;
+static constexpr int SPHERE_BLOCK = NPNP * SPHERE_BLOCK_LEV;
+
+struct SphereGlobal {
+
+  const ExecViewManaged<const Real *[2][2][NP][NP]> d;
+  const ExecViewManaged<const Real *[2][2][NP][NP]> dinv;
+  const ExecViewManaged<const Real [NP][NP]> dvv;
+  const ExecViewManaged<const Real *[NP][NP]> metdet;
+  const Real scale_factor_inv;
+
+  SphereGlobal(const SphereOperators &op):
+    d(op.m_d),
+    dinv(op.m_dinv),
+    dvv(op.dvv),
+    metdet(op.m_metdet),
+    scale_factor_inv(op.m_scale_factor_inv)
+  {}
+};
+
+using SphereBlockScratchView = Kokkos::View<
+  Real[SPHERE_BLOCK_LEV][NP][NP],
+  ExecSpace::scratch_memory_space,
+  Kokkos::MemoryTraits<Kokkos::Unmanaged>
+  >;
+
+using SphereBlockScratchSubview = Kokkos::Subview<
+  SphereBlockScratchView, int,
+  std::remove_const_t<decltype(Kokkos::ALL)>,
+  std::remove_const_t<decltype(Kokkos::ALL)>
+  >;
+
+struct SphereBlockOps;
+
+struct SphereBlockScratch {
+  SphereBlockScratchView v;
+  SphereBlockScratchSubview sv;
+  KOKKOS_INLINE_FUNCTION SphereBlockScratch(const SphereBlockOps &b);
+};
+
+struct SphereBlockOps {
+  const SphereGlobal &g;
+  const Team &t;
+  Real scale_factor_inv;
+  Real metdet;
+  Real rrdmd;
+  Real dinv[2][2];
+  Real dvvx[NP];
+  Real dvvy[NP];
+  int e,x,y,z;
+
+  KOKKOS_INLINE_FUNCTION SphereBlockOps(const SphereGlobal &sg, const Team &team):
+    g(sg),
+    t(team),
+    scale_factor_inv(g.scale_factor_inv),
+    rrdmd(0)
+  {
+    const int lr = t.league_rank();
+    e = lr / SPHERE_BLOCKS_PER_COL;
+    const int iw = lr % SPHERE_BLOCKS_PER_COL;
+    const int tr = t.team_rank();
+    const int ixy = tr / SPHERE_BLOCK_LEV;
+    x = ixy / NP;
+    y = ixy % NP;
+    const int dz = tr % SPHERE_BLOCK_LEV;
```

Comment:
This snippet below confuse me. I don't think team_rank() uniquely identifies a thread in a warp, but rather it identifies the warp as a whole. So I'm confused as of why this code can work correctly for all levels.

In Kokkos, team_rank() returns threadIdx.y, but a block (i.e., a team) is 2d.

Reply:
For `SphereBlockOps::policy`, the `team_rank()` does uniquely identify a thread, because the Kokkos "vector" dimension is one (no third argument in the constructor):
https://github.com/E3SM-Project/E3SM/blob/dd1ffec3501899234573866cdddb0c30c3c7449d/components/homme/src/share/cxx/SphereOperators.hpp#L1326
https://github.com/E3SM-Project/E3SM/blob/dd1ffec3501899234573866cdddb0c30c3c7449d/components/homme/src/share/cxx/SphereOperators.hpp#L1187
Other team policies do have nonzero vector lengths, so their `team_rank()` values do not uniquely identify a thread. This is the case for the existing code and for `SphereScanOps::policy` in the new code (three arguments in constructor):
https://github.com/E3SM-Project/E3SM/blob/dd1ffec3501899234573866cdddb0c30c3c7449d/components/homme/src/share/cxx/SphereOperators.hpp#L1442